contrib/memplot/memplot.nit

```diff
@@ -0,0 +1,198 @@
+# Program to transform `memory.log` files produced by `nitc --trace-memory` into a csv file
+module memplot
+
+import counter
+import template
+
+# An aggregated time-slice
+class Aggreg
+	# The start time of the slice
+	var time: Float
+
+	# Number of allocations
+	var acpt = new Counter[String]
+
+	# Number of allocated bytes
+	var asiz = new Counter[String]
+
+	# Number of deallocations
+	var dcpt = new Counter[String]
+
+	# Number of deallocated bytes
+	var dsiz = new Counter[String]
+
+	# Total number of allocations since the beginning
+	var cpttot = new Counter[String]
+
+	# Total number of allocated bytes since the beginning
+	var siztot = new Counter[String]
+
+	# Map of all the various counters.
+	var fields = new Map[String, Counter[String]]
+
+	init do
+		fields["acpt"] = acpt
+		fields["asiz"] = asiz
+		fields["dcpt"] = dcpt
+		fields["dsiz"] = dsiz
+		fields["cpttot"] = cpttot
+		fields["siztot"] = siztot
+	end
+end
+
+# Main class that does the job
+class MemProg
+	# The `memory.log` file.
+	var filepath: String
+
+	# The delay of an aggregation
+	var time_slice: Float
+
+	# Total number of events
+	var events = 0
+
+	# List a all time aggregations
+	var aggregs = new Array[Aggreg]
+
+	# Total number of allocations 
+	var cpttot = new Counter[String]
+
+	# Total number of allocated bytes
+	var siztot = new Counter[String]
+
+	# Parse the log file `filepath` and fill `aggregs`.
+	fun parse do
+		# Current lines (not yet put in an aggreg)
+		var lines = new Counter[String]
+
+		var time = 0.0
+		var dt = 100.0
+		dt = 100.0
+		for l in "memory.log".to_path.each_line do
+			if l[0] == '#' then
+				var t = l.substring_from(2).to_f
+
+				while t > time + dt do
+					aggreg(lines, time)
+					time += dt
+				end
+				#if time > 1000.0 then break
+				continue
+			end
+			events += 1
+			lines.inc(l)
+		end
+		aggreg(lines, time)
+	end
+
+	# Create and register a new aggregation
+	fun aggreg(lines: Counter[String], t1: Float) do
+		var aggreg = new Aggreg(t1)
+		aggregs.add aggreg
+		print "events:{events} aggregs:{aggregs.length} {t1}ms"
+
+		# Process each distinct line
+		for l, v in lines do
+			var a = l.split('\t')
+			if a.length != 3 then
+				print "Error {a.length}. {l}"
+				continue
+			end
+			var c = a[0]
+			var s = a[1].to_i
+			var e = a[2]
+			if c == "+" then
+				aggreg.acpt[e] += v
+				aggreg.asiz[e] += v * s
+			else if c == "-" then
+				aggreg.dcpt[e] += v
+				aggreg.dsiz[e] += v * s
+			else abort
+		end
+		lines.clear
+
+		# Sum all information
+		for e, v in aggreg.acpt do cpttot[e] += v
+		for e, v in aggreg.asiz do siztot[e] += v
+		for e, v in aggreg.dcpt do cpttot[e] -= v
+		for e, v in aggreg.dsiz do siztot[e] -= v
+
+		# Copy current total
+		aggreg.cpttot.add_all cpttot
+		aggreg.siztot.add_all siztot
+
+		cpttot.print_elements(2)
+	end
+
+	# Generate a *long* CVS file, to use with statistical tools
+	fun tolong: Writable
+	do
+		var res = new Template
+
+		# Write the header
+		res.add "time, class"
```

Comment:
CSV files are whitespace-sensitive. For example, a RFC4180-compliant parser should see this like that (note the space before `class`):
```
["time", " class"]
```